server/src/test/java/org/elasticsearch/ingest/IngestLocalShardSearcherTests.java

```diff
@@ -0,0 +1,235 @@
+            client().index(new IndexRequest("reference-index").id("1").source("{}", XContentType.JSON)).actionGet();
+            client().admin().indices().refresh(new RefreshRequest("reference-index")).actionGet();
+
+            PutPipelineRequest putPipelineRequest = new PutPipelineRequest("my-pipeline", createPipelineSource(), XContentType.JSON);
+            client().admin().cluster().putPipeline(putPipelineRequest).get();
+
+            client().index(new IndexRequest("my-index").id("1").source("{}", XContentType.JSON).setPipeline("my-pipeline")).actionGet();
+            client().admin().indices().refresh(new RefreshRequest("my-index")).actionGet();
+
+            Map<String, Object> result = client().get(new GetRequest("my-index", "1")).actionGet().getSourceAsMap();
+            assertThat(result.size(), equalTo(1));
+            assertThat(result.get("id"), equalTo("1"));
+            assertThat(SEARCH_WRAPPER_APPLIED.get(), is(true));
+        } finally {
+            SET_SEARCH_WRAPPER.set(false);
+            SEARCH_WRAPPER_APPLIED.set(false);
+        }
+    }
+
+    private static BytesReference createPipelineSource() throws IOException {
+        return BytesReference.bytes(jsonBuilder().startObject()
+            .startArray("processors")
+            .startObject()
+            .startObject(TestProcessor.NAME)
+            .endObject()
+            .endObject()
+            .endArray()
+            .endObject());
+    }
+
+    public static class TestPlugin extends Plugin implements IngestPlugin {
+
+        @Override
+        public Map<String, Processor.Factory> getProcessors(Processor.Parameters parameters) {
+            return Collections.singletonMap(TestProcessor.NAME, new TestProcessor.Factory(parameters.localShardSearcher));
+        }
+
+        @Override
+        public void onIndexModule(IndexModule indexModule) {
+            if (SET_SEARCH_WRAPPER.get()) {
+                indexModule.setSearcherWrapper(indexService -> new IndexSearcherWrapper() {
+
+                    @Override
+                    protected DirectoryReader wrap(DirectoryReader reader) throws IOException {
+                        SEARCH_WRAPPER_APPLIED.set(true);
+                        return super.wrap(reader);
+                    }
+                });
+            }
+        }
+    }
+
+    static class TestProcessor extends AbstractProcessor {
+
+        static final String NAME = "test_processor";
+
+        private final Function<String, Engine.Searcher> localShardSearcher;
+
+        TestProcessor(String tag, Function<String, Engine.Searcher> localShardSearcher) {
+            super(tag);
+            this.localShardSearcher = localShardSearcher;
+        }
+
+        @Override
+        public IngestDocument execute(IngestDocument ingestDocument) throws Exception {
+            String indexExpression = "reference-index";
+            try (Engine.Searcher engineSearcher = localShardSearcher.apply(indexExpression)) {
```

Comment:
Asserting that the wrapper has been called doesn't feel enough to me, I'd like to have a custom directory reader that the wrapper uses, and assert here that we get the expected implementation.

Reply:
I've adjusted this test: https://github.com/elastic/elasticsearch/pull/41010/commits/fc939ef1b37725dfc9b9fa23d8559364539d0057